<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 <HEAD>
   <TITLE> thread-safety definition
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20thread-safety%20definition&In-Reply-To=%3Cullo2koc1.fsf%40boost-consulting.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="000295.html">
   <LINK REL="Next"  HREF="000303.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>thread-safety definition</H1>
    <B>David Abrahams</B> 
    <A HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20thread-safety%20definition&In-Reply-To=%3Cullo2koc1.fsf%40boost-consulting.com%3E"
       TITLE="thread-safety definition">dave at boost-consulting.com
       </A><BR>
    <I>Wed Jan 21 02:49:50 UTC 2004</I>
    <P><UL>
        <LI>Previous message: <A HREF="000295.html">[c++-pthreads] Re: thread-safety definition
</A></li>
        <LI>Next message: <A HREF="000303.html">[c++-pthreads] Re: thread-safety definition
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#300">[ date ]</a>
              <a href="thread.html#300">[ thread ]</a>
              <a href="subject.html#300">[ subject ]</a>
              <a href="author.html#300">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Dave Butenhof &lt;<A HREF="http://sourcerytools.com/cgi-bin/mailman/listinfo/c++-pthreads">David.Butenhof at hp.com</A>&gt; writes:

&gt;&gt;<i>My point is that one could just as well (and perhaps more justifiably)
</I>&gt;&gt;<i>claim that it's evil to throw any exception that can't be dealt with
</I>&gt;&gt;<i>via a catch(...) block that doesn't rethrow, where that block was
</I>&gt;&gt;<i>designed to handle, say, bad_alloc.**
</I>&gt;&gt;<i>  
</I>&gt;&gt;<i>
</I>&gt;<i> The real issue isn't so much catching without a re-throw, but catching
</I>&gt;<i> and dropping arbitrary exceptions. If you don't claim to know how to
</I>&gt;<i> finalize a bad_alloc error, what business do you have catching it?
</I>&gt;<i> (And since catch(...) is anonymous, even if you could handle it you
</I>&gt;<i> can't identify it.) I'm not convinced there's any rational excuse for
</I>&gt;<i> finalizing bad_alloc unless you know how to free up some memory to
</I>&gt;<i> &quot;cure&quot; the problem that caused it. (Nor would there be any particular
</I>&gt;<i> benefit.) If you can't deal with it, you need to let it propagate to
</I>&gt;<i> someone who can; or to terminate the process.
</I>
Let me give you one example.  I used to write desktop application
software.  This software had an undo facility.  The undo mechanism
would store a copy of the portion of the document's original data that
was about to be changed before any change was made.  This arrangement
also allowed any failing operation to be rolled back.  Sometimes the
user would try to do something that would consume more than the
available memory, and sometimes the user would try to do something
that ate up the last bit of disk space, and sometimes an operation
would fail for other reasons.  I could always bring the user back to
the main event loop and fully recover her document's data so she could
try to save it.  The only reason the type of the exception mattered
was that it allowed me to give the user a good error report.  If there
was an exception I couldn't recognize, it would always be better to
say &quot;something I can't identify went wrong; you might want to save
under a different name&quot; than to allow the program to terminate
suddenly, and *definitely* drop the document's data on the floor.  Of
course, I never saw an exception I couldn't identify, but termination
is not always the right answer.

Another real-life example someone once gave me was that of a stage
lighting controller.  It was always better to plow ahead and hope the
issue wasn't serious than to have the stage go suddenly black.

There are surely places where finalizing a catch(...) is wrong, but
there's no blanket rule that applies to all applications.

-- 
Dave Abrahams
Boost Consulting
www.boost-consulting.com


</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="000295.html">[c++-pthreads] Re: thread-safety definition
</A></li>
	<LI>Next message: <A HREF="000303.html">[c++-pthreads] Re: thread-safety definition
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#300">[ date ]</a>
              <a href="thread.html#300">[ thread ]</a>
              <a href="subject.html#300">[ subject ]</a>
              <a href="author.html#300">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://sourcerytools.com/cgi-bin/mailman/listinfo/c++-pthreads">More information about the c++-pthreads
mailing list</a><br>
</body></html>
